error: failed to run custom build command for `mupdf-sys v0.0.3`

[dependencies]
mupdf = "0.0.3"


Compilation error：
error: failed to run custom build command for `mupdf-sys v0.0.3`  

Whether mupdf needs to be installed first, how to install it

I had the same problem, using Windows 10.
Noticing that line 43 of `mupdf-sys-0.0.3\build.rs` invokes the GNU `make` utility, i installed that via the [Chocolatey](https://chocolatey.org/install) package manager; now the error is
```
PS C:\Users\user\Documents\GitHub\ebook-juggler\ebook-juggler> cargo build
   Compiling mupdf-sys v0.0.3
error: failed to run custom build command for `mupdf-sys v0.0.3`

Caused by:
  process didn't exit successfully: `C:\Users\user\Documents\GitHub\ebook-juggler\ebook-juggler\target\debug\build\mupdf-sys-cb0d3627e8183403\build-script-build` (exit code: 101)
  --- stdout
  cargo:rerun-if-changed=wrapper.h
  cargo:rerun-if-changed=wrapper.c
  mkdir -p C:\Users\user\Documents\GitHub\ebook-juggler\ebook-juggler\target\debug\build\mupdf-sys-4950eff510008997\out/source/fitz/ ; cc -ffunction-sections -fdata-sections -pipe -g -DNOTO_SMALL -DNO_CJK -Iinclude -MMD -MP -o C:\Users\user\Documents\GitHub\ebook-juggler\ebook-juggler\target\debug\build\mupdf-sys-4950eff510008997\out/source/fitz/archive.o -c source/fitz/archive.c -Wall -Wdeclaration-after-statement -Iscripts/freetype -Ithirdparty/freetype/include -Ithirdparty/harfbuzz/src -Iscripts/libjpeg -Ithirdparty/libjpeg -Ithirdparty/lcms2/include -DHAVE_LCMS2MT -Ithirdparty/mujs -Ithirdparty/zlib -Ithirdparty/jbig2dec -Ithirdparty/openjpeg/src/lib/openjp2 -DOPJ_STATIC -DOPJ_HAVE_STDINT_H -DOPJ_HAVE_INTTYPES_H -DMUTEX_pthread=0

  --- stderr
  process_begin: CreateProcess(NULL, uname, ...) failed.
  Makerules:3: pipe: No error
  File not found - *.[ch]
  The syntax of the command is incorrect.
  make: *** [Makefile:101: C:\Users\user\Documents\GitHub\ebook-juggler\ebook-juggler\target\debug\build\mupdf-sys-4950eff510008997\out/source/fitz/archive.o] Error 1
  thread 'main' panicked at '""', C:\Users\user\.cargo\registry\src\github.com-1ecc6299db9ec823\mupdf-sys-0.0.3\build.rs:45:9
```
MuPDF can definitely be compiled for Windows as they offer [pre-built binaries on their site](https://mupdf.com/downloads/index.html) (although only as apps, not libraries for linking to, unfortunately).

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -14,11 +14,6 @@ jobs:
             auth_header="$(git config --local --get http.https://github.com/.extraheader)"
             git submodule sync --recursive
             git -c "http.extraheader=$auth_header" -c protocol.version=2 submodule update --init --force --recursive --depth=1
-      - name: Cache cargo build
-        uses: actions/cache@v1
-        with:
-          path: target
-          key: ${{ runner.os }}-cargo-build-target-${{ hashFiles('**/Cargo.lock') }}
       - uses: actions-rs/toolchain@v1
         with:
           profile: minimal
@@ -33,7 +28,7 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        os: [ubuntu-latest, macos-latest]
+        os: [ubuntu-latest, macos-latest, windows-latest]
     steps:
       - uses: actions/checkout@v2
       - name: Checkout submodules
@@ -47,7 +42,17 @@ jobs:
           profile: minimal
           toolchain: stable
           override: true
+      - name: Install LLVM
+        if: matrix.os == 'windows-latest'
+        run: choco install -y llvm
       - uses: actions-rs/cargo@v1
+        if: matrix.os == 'windows-latest'
+        env:
+          LIBCLANG_PATH: "C:\\Program Files\\LLVM\\bin"
+        with:
+          command: test
+      - uses: actions-rs/cargo@v1
+        if: matrix.os != 'windows-latest'
         with:
           command: test
 

--- a/mupdf-sys/build.rs
+++ b/mupdf-sys/build.rs
@@ -1,7 +1,7 @@
 use std::env;
 use std::fs;
-use std::path::PathBuf;
-use std::process::{Command, Stdio};
+use std::path::{Path, PathBuf};
+use std::process::Stdio;
 
 // see https://github.com/ArtifexSoftware/mupdf/blob/master/source/fitz/noto.c
 const SKIP_FONTS: [&str; 6] = [
@@ -24,18 +24,45 @@ fn fail_on_empty_directory(name: &str) {
     }
 }
 
-fn main() {
-    fail_on_empty_directory("mupdf");
-    println!("cargo:rerun-if-changed=wrapper.h");
-    println!("cargo:rerun-if-changed=wrapper.c");
+macro_rules! t {
+    ($e:expr) => {
+        match $e {
+            Ok(n) => n,
+            Err(e) => panic!("\n{} failed with {}\n", stringify!($e), e),
+        }
+    };
+}
+
+fn cp_r(dir: &Path, dest: &Path) {
+    for entry in t!(fs::read_dir(dir)) {
+        let entry = t!(entry);
+        let path = entry.path();
+        let dst = dest.join(path.file_name().expect("Failed to get filename of path"));
+        if t!(fs::metadata(&path)).is_file() {
+            t!(fs::copy(path, dst));
+        } else {
+            t!(fs::create_dir_all(&dst));
+            cp_r(&path, &dst);
+        }
+    }
+}
+
+#[cfg(not(target_env = "msvc"))]
+fn build_libmupdf() {
+    use std::process::Command;
 
     let profile = match &*env::var("PROFILE").unwrap_or("debug".to_owned()) {
         "bench" | "release" => "release",
         _ => "debug",
     };
-    let out_dir = env::var("OUT_DIR").unwrap();
+    let out_dir = PathBuf::from(env::var("OUT_DIR").unwrap());
+    let build_dir = out_dir.join("build");
+    t!(fs::create_dir_all(&build_dir));
+
     let current_dir = env::current_dir().unwrap();
-    let mupdf_dir = current_dir.join("mupdf");
+    let mupdf_src_dir = current_dir.join("mupdf");
+    cp_r(&mupdf_src_dir, &build_dir);
+
     // see https://github.com/ArtifexSoftware/mupdf/blob/master/include/mupdf/fitz/config.h
     let xcflags = vec![
         #[cfg(not(feature = "xps"))]
@@ -61,7 +88,7 @@ fn main() {
     let make_flags = vec![
         "libs".to_owned(),
         format!("build={}", profile),
-        format!("OUT={}", out_dir),
+        format!("OUT={}", build_dir.display()),
         #[cfg(feature = "sys-lib")]
         "USE_SYSTEM_LIBS=yes".to_owned(),
         "HAVE_X11=no".to_owned(),
@@ -88,19 +115,103 @@ fn main() {
 
     let output = Command::new("make")
         .args(&make_flags)
-        .current_dir(mupdf_dir)
+        .current_dir(&build_dir)
         .stdout(Stdio::inherit())
         .stderr(Stdio::inherit())
         .output()
         .expect("make failed");
-    if output.status.code() != Some(0) {
-        panic!("{:?}", String::from_utf8(output.stdout).unwrap());
+    if !output.status.success() {
+        let err = String::from_utf8_lossy(&output.stderr);
+        let out = String::from_utf8_lossy(&output.stdout);
+        panic!("Build error:\nSTDERR:{}\nSTDOUT:{}", err, out);
     }
-    println!("cargo:rustc-link-search=native={}", out_dir);
+    println!("cargo:rustc-link-search=native={}", build_dir.display());
     println!("cargo:rustc-link-lib=static=mupdf");
     // println!("cargo:rustc-link-lib=static=mupdf-pkcs7");
     println!("cargo:rustc-link-lib=static=mupdf-third");
     // println!("cargo:rustc-link-lib=static=mupdf-threads");
+}
+
+#[cfg(target_env = "msvc")]
+fn build_libmupdf() {
+    let target = env::var("TARGET").expect("TARGET not found in environment");
+    let msvc_platform = if target.contains("x86_64") {
+        "x64"
+    } else {
+        "Win32"
+    };
+    let profile = match &*env::var("PROFILE").unwrap_or("debug".to_owned()) {
+        "bench" | "release" => "Release",
+        _ => "Debug",
+    };
+    let out_dir = PathBuf::from(env::var("OUT_DIR").unwrap());
+    let build_dir = out_dir.join("build");
+    t!(fs::create_dir_all(&build_dir));
+
+    let current_dir = env::current_dir().unwrap();
+    let mupdf_src_dir = current_dir.join("mupdf");
+    cp_r(&mupdf_src_dir, &build_dir);
+
+    let devenv = cc::windows_registry::find(target.as_str(), "devenv.exe");
+    if let Some(mut devenv) = devenv {
+        let d = devenv
+            .args(&["/upgrade", "platform\\win32\\mupdf.sln"])
+            .current_dir(&build_dir)
+            .stdout(Stdio::inherit())
+            .stderr(Stdio::inherit())
+            .output()
+            .expect("failed to run devenv. Do you have it installed?");
+        if !d.status.success() {
+            let err = String::from_utf8_lossy(&d.stderr);
+            let out = String::from_utf8_lossy(&d.stdout);
+            eprintln!("Upgrade error:\nSTDERR:{}\nSTDOUT:{}", err, out);
+        }
+        let d = cc::windows_registry::find(target.as_str(), "devenv.exe")
+            .unwrap()
+            .args(&[
+                "platform\\win32\\mupdf.sln",
+                "/build",
+                &format!("{}|{}", profile, msvc_platform),
+                "/project",
+                "libmupdf",
+            ])
+            .current_dir(&build_dir)
+            .stdout(Stdio::inherit())
+            .stderr(Stdio::inherit())
+            .output()
+            .expect("failed to run devenv. Do you have it installed?");
+        if !d.status.success() {
+            let err = String::from_utf8_lossy(&d.stderr);
+            let out = String::from_utf8_lossy(&d.stdout);
+            panic!("Build error:\nSTDERR:{}\nSTDOUT:{}", err, out);
+        }
+        if msvc_platform == "x64" {
+            println!(
+                "cargo:rustc-link-search=native={}/platform/win32/{}/{}",
+                build_dir.display(),
+                msvc_platform,
+                profile
+            );
+        } else {
+            println!(
+                "cargo:rustc-link-search=native={}/platform/win32/{}",
+                build_dir.display(),
+                profile
+            );
+        }
+        println!("cargo:rustc-link-lib=dylib=libmupdf");
+        println!("cargo:rustc-link-lib=dylib=libthirdparty");
+    } else {
+        eprintln!("failed to find devenv. Do you have it installed?");
+    }
+}
+
+fn main() {
+    fail_on_empty_directory("mupdf");
+    println!("cargo:rerun-if-changed=wrapper.h");
+    println!("cargo:rerun-if-changed=wrapper.c");
+
+    build_libmupdf();
 
     let mut build = cc::Build::new();
     build.file("wrapper.c");

--- a/mupdf-sys/wrapper.c
+++ b/mupdf-sys/wrapper.c
@@ -488,7 +488,7 @@ fz_stroke_state *mupdf_default_stroke_state(fz_context *ctx)
 }
 
 fz_stroke_state *mupdf_new_stroke_state(
-    fz_context *ctx, uint32_t start_cap, uint32_t dash_cap, uint32_t end_cap, uint32_t line_join, float line_width,
+    fz_context *ctx, fz_linecap start_cap, fz_linecap dash_cap, fz_linecap end_cap, fz_linejoin line_join, float line_width,
     float miter_limit, float dash_phase, const float dash[], int dash_len, mupdf_error_t **errptr)
 {
     fz_stroke_state *stroke = NULL;

--- a/src/pdf/document.rs
+++ b/src/pdf/document.rs
@@ -27,7 +27,8 @@ bitflags! {
 }
 
 #[derive(Debug, Copy, Clone, PartialEq, TryFromPrimitive)]
-#[repr(u32)]
+#[cfg_attr(target_env = "msvc", repr(i32))]
+#[cfg_attr(not(target_env = "msvc"), repr(u32))]
 pub enum Encryption {
     Aes128 = PDF_ENCRYPT_AES_128,
     Aes256 = PDF_ENCRYPT_AES_256,
@@ -168,6 +169,12 @@ impl PdfWriteOptions {
         self
     }
 
+    #[cfg(target_env = "msvc")]
+    pub fn encryption(&self) -> Encryption {
+        Encryption::try_from(self.inner.do_encrypt).unwrap()
+    }
+
+    #[cfg(not(target_env = "msvc"))]
     pub fn encryption(&self) -> Encryption {
         Encryption::try_from(self.inner.do_encrypt as u32).unwrap()
     }

--- a/src/pdf/document.rs
+++ b/src/pdf/document.rs
@@ -27,16 +27,15 @@ bitflags! {
 }
 
 #[derive(Debug, Copy, Clone, PartialEq, TryFromPrimitive)]
-#[cfg_attr(target_env = "msvc", repr(i32))]
-#[cfg_attr(not(target_env = "msvc"), repr(u32))]
+#[repr(u32)]
 pub enum Encryption {
-    Aes128 = PDF_ENCRYPT_AES_128,
-    Aes256 = PDF_ENCRYPT_AES_256,
-    Rc4_40 = PDF_ENCRYPT_RC4_40,
-    Rc4_128 = PDF_ENCRYPT_RC4_128,
-    Keep = PDF_ENCRYPT_KEEP,
-    None = PDF_ENCRYPT_NONE,
-    Unknown = PDF_ENCRYPT_UNKNOWN,
+    Aes128 = PDF_ENCRYPT_AES_128 as u32,
+    Aes256 = PDF_ENCRYPT_AES_256 as u32,
+    Rc4_40 = PDF_ENCRYPT_RC4_40 as u32,
+    Rc4_128 = PDF_ENCRYPT_RC4_128 as u32,
+    Keep = PDF_ENCRYPT_KEEP as u32,
+    None = PDF_ENCRYPT_NONE as u32,
+    Unknown = PDF_ENCRYPT_UNKNOWN as u32,
 }
 
 impl Default for Encryption {
@@ -169,12 +168,6 @@ impl PdfWriteOptions {
         self
     }
 
-    #[cfg(target_env = "msvc")]
-    pub fn encryption(&self) -> Encryption {
-        Encryption::try_from(self.inner.do_encrypt).unwrap()
-    }
-
-    #[cfg(not(target_env = "msvc"))]
     pub fn encryption(&self) -> Encryption {
         Encryption::try_from(self.inner.do_encrypt as u32).unwrap()
     }
@@ -438,7 +431,7 @@ impl PdfDocument {
         Ok(())
     }
 
-    pub fn is_js_supported(&mut self) -> Result<bool, Error> {
+    pub fn is_js_supported(&self) -> Result<bool, Error> {
         let supported = unsafe { ffi_try!(mupdf_pdf_js_supported(context(), self.inner)) };
         Ok(supported)
     }

--- a/src/stroke_state.rs
+++ b/src/stroke_state.rs
@@ -6,7 +6,8 @@ use num_enum::TryFromPrimitive;
 use crate::{context, Error};
 
 #[derive(Debug, Clone, Copy, PartialEq, TryFromPrimitive)]
-#[repr(u32)]
+#[cfg_attr(target_env = "msvc", repr(i32))]
+#[cfg_attr(not(target_env = "msvc"), repr(u32))]
 pub enum LineCap {
     Butt = 0,
     Round = 1,
@@ -21,7 +22,8 @@ impl Default for LineCap {
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, TryFromPrimitive)]
-#[repr(u32)]
+#[cfg_attr(target_env = "msvc", repr(i32))]
+#[cfg_attr(not(target_env = "msvc"), repr(u32))]
 pub enum LineJoin {
     Miter = 0,
     Round = 1,

--- a/src/stroke_state.rs
+++ b/src/stroke_state.rs
@@ -6,8 +6,7 @@ use num_enum::TryFromPrimitive;
 use crate::{context, Error};
 
 #[derive(Debug, Clone, Copy, PartialEq, TryFromPrimitive)]
-#[cfg_attr(target_env = "msvc", repr(i32))]
-#[cfg_attr(not(target_env = "msvc"), repr(u32))]
+#[repr(u32)]
 pub enum LineCap {
     Butt = 0,
     Round = 1,
@@ -22,8 +21,7 @@ impl Default for LineCap {
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, TryFromPrimitive)]
-#[cfg_attr(target_env = "msvc", repr(i32))]
-#[cfg_attr(not(target_env = "msvc"), repr(u32))]
+#[repr(u32)]
 pub enum LineJoin {
     Miter = 0,
     Round = 1,
@@ -93,19 +91,19 @@ impl StrokeState {
     }
 
     pub fn start_cap(&self) -> LineCap {
-        LineCap::try_from(unsafe { (*self.inner).start_cap }).unwrap()
+        LineCap::try_from(unsafe { (*self.inner).start_cap as u32 }).unwrap()
     }
 
     pub fn dash_cap(&self) -> LineCap {
-        LineCap::try_from(unsafe { (*self.inner).dash_cap }).unwrap()
+        LineCap::try_from(unsafe { (*self.inner).dash_cap as u32 }).unwrap()
     }
 
     pub fn end_cap(&self) -> LineCap {
-        LineCap::try_from(unsafe { (*self.inner).end_cap }).unwrap()
+        LineCap::try_from(unsafe { (*self.inner).end_cap as u32 }).unwrap()
     }
 
     pub fn line_join(&self) -> LineJoin {
-        LineJoin::try_from(unsafe { (*self.inner).linejoin }).unwrap()
+        LineJoin::try_from(unsafe { (*self.inner).linejoin as u32 }).unwrap()
     }
 
     pub fn line_width(&self) -> f32 {

--- a/src/text.rs
+++ b/src/text.rs
@@ -47,26 +47,24 @@ impl Drop for Text {
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, TryFromPrimitive)]
-#[cfg_attr(target_env = "msvc", repr(i32))]
-#[cfg_attr(not(target_env = "msvc"), repr(u32))]
+#[repr(u32)]
 pub enum BidiDirection {
-    Ltr = fz_bidi_direction_FZ_BIDI_LTR,
-    Neutral = fz_bidi_direction_FZ_BIDI_NEUTRAL,
-    Rtl = fz_bidi_direction_FZ_BIDI_RTL,
+    Ltr = fz_bidi_direction_FZ_BIDI_LTR as u32,
+    Neutral = fz_bidi_direction_FZ_BIDI_NEUTRAL as u32,
+    Rtl = fz_bidi_direction_FZ_BIDI_RTL as u32,
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, TryFromPrimitive)]
-#[cfg_attr(target_env = "msvc", repr(i32))]
-#[cfg_attr(not(target_env = "msvc"), repr(u32))]
+#[repr(u32)]
 pub enum Language {
-    Unset = fz_text_language_FZ_LANG_UNSET,
-    Ja = fz_text_language_FZ_LANG_ja,
-    Ko = fz_text_language_FZ_LANG_ko,
-    Ur = fz_text_language_FZ_LANG_ur,
-    Urd = fz_text_language_FZ_LANG_urd,
-    Zh = fz_text_language_FZ_LANG_zh,
-    ZhHans = fz_text_language_FZ_LANG_zh_Hans,
-    ZhHant = fz_text_language_FZ_LANG_zh_Hant,
+    Unset = fz_text_language_FZ_LANG_UNSET as u32,
+    Ja = fz_text_language_FZ_LANG_ja as u32,
+    Ko = fz_text_language_FZ_LANG_ko as u32,
+    Ur = fz_text_language_FZ_LANG_ur as u32,
+    Urd = fz_text_language_FZ_LANG_urd as u32,
+    Zh = fz_text_language_FZ_LANG_zh as u32,
+    ZhHans = fz_text_language_FZ_LANG_zh_Hans as u32,
+    ZhHant = fz_text_language_FZ_LANG_zh_Hant as u32,
 }
 
 #[derive(Debug)]
@@ -105,12 +103,6 @@ impl TextSpan {
         unsafe { (*self.inner).set_bidi_level(bidi_level) }
     }
 
-    #[cfg(target_env = "msvc")]
-    pub fn markup_dir(&self) -> BidiDirection {
-        unsafe { ((*self.inner).markup_dir() as i32).try_into().unwrap() }
-    }
-
-    #[cfg(not(target_env = "msvc"))]
     pub fn markup_dir(&self) -> BidiDirection {
         unsafe { (*self.inner).markup_dir().try_into().unwrap() }
     }
@@ -119,12 +111,6 @@ impl TextSpan {
         unsafe { (*self.inner).set_markup_dir(dir as _) }
     }
 
-    #[cfg(target_env = "msvc")]
-    pub fn language(&self) -> Language {
-        unsafe { ((*self.inner).language() as i32).try_into().unwrap() }
-    }
-
-    #[cfg(not(target_env = "msvc"))]
     pub fn language(&self) -> Language {
         unsafe { (*self.inner).language().try_into().unwrap() }
     }

--- a/src/text.rs
+++ b/src/text.rs
@@ -47,7 +47,8 @@ impl Drop for Text {
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, TryFromPrimitive)]
-#[repr(u32)]
+#[cfg_attr(target_env = "msvc", repr(i32))]
+#[cfg_attr(not(target_env = "msvc"), repr(u32))]
 pub enum BidiDirection {
     Ltr = fz_bidi_direction_FZ_BIDI_LTR,
     Neutral = fz_bidi_direction_FZ_BIDI_NEUTRAL,
@@ -55,7 +56,8 @@ pub enum BidiDirection {
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, TryFromPrimitive)]
-#[repr(u32)]
+#[cfg_attr(target_env = "msvc", repr(i32))]
+#[cfg_attr(not(target_env = "msvc"), repr(u32))]
 pub enum Language {
     Unset = fz_text_language_FZ_LANG_UNSET,
     Ja = fz_text_language_FZ_LANG_ja,
@@ -103,6 +105,12 @@ impl TextSpan {
         unsafe { (*self.inner).set_bidi_level(bidi_level) }
     }
 
+    #[cfg(target_env = "msvc")]
+    pub fn markup_dir(&self) -> BidiDirection {
+        unsafe { ((*self.inner).markup_dir() as i32).try_into().unwrap() }
+    }
+
+    #[cfg(not(target_env = "msvc"))]
     pub fn markup_dir(&self) -> BidiDirection {
         unsafe { (*self.inner).markup_dir().try_into().unwrap() }
     }
@@ -111,6 +119,12 @@ impl TextSpan {
         unsafe { (*self.inner).set_markup_dir(dir as _) }
     }
 
+    #[cfg(target_env = "msvc")]
+    pub fn language(&self) -> Language {
+        unsafe { ((*self.inner).language() as i32).try_into().unwrap() }
+    }
+
+    #[cfg(not(target_env = "msvc"))]
     pub fn language(&self) -> Language {
         unsafe { (*self.inner).language().try_into().unwrap() }
     }

--- a/src/text_page.rs
+++ b/src/text_page.rs
@@ -89,11 +89,10 @@ impl Drop for TextPage {
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, TryFromPrimitive)]
-#[cfg_attr(target_env = "msvc", repr(i32))]
-#[cfg_attr(not(target_env = "msvc"), repr(u32))]
+#[repr(u32)]
 pub enum TextBlockType {
-    Text = FZ_STEXT_BLOCK_TEXT,
-    Image = FZ_STEXT_BLOCK_IMAGE,
+    Text = FZ_STEXT_BLOCK_TEXT as u32,
+    Image = FZ_STEXT_BLOCK_IMAGE as u32,
 }
 
 #[derive(Debug)]
@@ -102,12 +101,6 @@ pub struct TextBlock {
 }
 
 impl TextBlock {
-    #[cfg(target_env = "msvc")]
-    pub fn r#type(&self) -> TextBlockType {
-        unsafe { ((*self.inner).type_).try_into().unwrap() }
-    }
-
-    #[cfg(not(target_env = "msvc"))]
     pub fn r#type(&self) -> TextBlockType {
         unsafe { ((*self.inner).type_ as u32).try_into().unwrap() }
     }

--- a/src/text_page.rs
+++ b/src/text_page.rs
@@ -89,7 +89,8 @@ impl Drop for TextPage {
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, TryFromPrimitive)]
-#[repr(u32)]
+#[cfg_attr(target_env = "msvc", repr(i32))]
+#[cfg_attr(not(target_env = "msvc"), repr(u32))]
 pub enum TextBlockType {
     Text = FZ_STEXT_BLOCK_TEXT,
     Image = FZ_STEXT_BLOCK_IMAGE,
@@ -101,6 +102,12 @@ pub struct TextBlock {
 }
 
 impl TextBlock {
+    #[cfg(target_env = "msvc")]
+    pub fn r#type(&self) -> TextBlockType {
+        unsafe { ((*self.inner).type_).try_into().unwrap() }
+    }
+
+    #[cfg(not(target_env = "msvc"))]
     pub fn r#type(&self) -> TextBlockType {
         unsafe { ((*self.inner).type_ as u32).try_into().unwrap() }
     }
@@ -111,7 +118,7 @@ impl TextBlock {
 
     pub fn lines(&self) -> TextLineIter {
         unsafe {
-            if (*self.inner).type_ as u32 == FZ_STEXT_BLOCK_TEXT {
+            if (*self.inner).type_ == FZ_STEXT_BLOCK_TEXT as _ {
                 return TextLineIter {
                     next: (*self.inner).u.t.first_line,
                 };
@@ -124,7 +131,7 @@ impl TextBlock {
 
     pub fn ctm(&self) -> Option<Matrix> {
         unsafe {
-            if (*self.inner).type_ as u32 == FZ_STEXT_BLOCK_IMAGE {
+            if (*self.inner).type_ == FZ_STEXT_BLOCK_IMAGE as _ {
                 return Some((*self.inner).u.i.transform.into());
             }
         }
@@ -133,7 +140,7 @@ impl TextBlock {
 
     pub fn image(&self) -> Option<Image> {
         unsafe {
-            if (*self.inner).type_ as u32 == FZ_STEXT_BLOCK_IMAGE {
+            if (*self.inner).type_ == FZ_STEXT_BLOCK_IMAGE as _ {
                 let inner = (*self.inner).u.i.image;
                 fz_keep_image(context(), inner);
                 return Some(Image::from_raw(inner));

